File: android/src/main/kotlin/com/loopit/minis/imgedit/AssetCatalog.kt
package com.loopit.minis.imgedit

import android.content.Context
import android.util.Log
import org.json.JSONObject

/**
 * Enumerates bundled LUTs, sticker packs, and fonts.
 *
 * Reads:
 *   - `android/src/main/assets/luts/<id>.cube` — listed by stem
 *   - `android/src/main/assets/stickers/<pack_id>/manifest.json`
 *   - `android/src/main/assets/fonts/<family>.ttf`
 *
 * Resolution uses the [Context] supplied to [bind]; the router holds the
 * context for the lifetime of the editor session. If [bind] hasn't been
 * called yet, calls return safe defaults so the channel surface keeps
 * working.
 */
object AssetCatalog {

    private const val TAG = "MinisImgEditAssets"
    private var context: Context? = null

    fun bind(ctx: Context?) {
        context = ctx?.applicationContext
    }

    fun filters(): List<Map<String, Any?>> {
        val ctx = context ?: return listOf(stub("neutral", "Neutral"))
        return try {
            val names = ctx.assets.list("luts").orEmpty()
            val out = mutableListOf<Map<String, Any?>>()
            out += stub("neutral", "Neutral")
            for (name in names) {
                if (!name.endsWith(".cube")) continue
                val id = name.removeSuffix(".cube")
                out += mapOf(
                    "id" to id,
                    "label" to id.replaceFirstChar { it.uppercase() },
                    "lutPath" to "asset:luts/$name",
                )
            }
            out
        } catch (t: Throwable) {
            Log.w(TAG, "filters() failed: ${t.message}")
            listOf(stub("neutral", "Neutral"))
        }
    }

    fun stickerPacks(): List<Map<String, Any?>> {
        val ctx = context ?: return emptyList()
        return try {
            val packs = ctx.assets.list("stickers").orEmpty()
            val out = mutableListOf<Map<String, Any?>>()
            for (pack in packs) {
                val manifest = try {
                    ctx.assets.open("stickers/$pack/manifest.json").use { it.readBytes() }
                } catch (_: Throwable) { null }
                val json = manifest?.let { JSONObject(String(it)) }
                val label = json?.optString("label") ?: pack
                val itemsArr = json?.optJSONArray("items")
                val items = mutableListOf<Map<String, Any?>>()
                if (itemsArr != null) {
                    for (i in 0 until itemsArr.length()) {
                        val o = itemsArr.optJSONObject(i) ?: continue
                        items += mapOf(
                            "id" to o.optString("id"),
                            "w" to o.optInt("w"),
                            "h" to o.optInt("h"),
                            "path" to "asset:stickers/$pack/${o.optString("file")}",
                        )
                    }
                }
                out += mapOf(
                    "packId" to pack,
                    "displayName" to label,
                    "items" to items,
                )
            }
            out
        } catch (t: Throwable) {
            Log.w(TAG, "stickerPacks() failed: ${t.message}")
            emptyList()
        }
    }

    fun fonts(): List<Map<String, Any?>> {
        val ctx = context ?: return listOf(mapOf("family" to "system", "weights" to listOf("regular")))
        return try {
            val names = ctx.assets.list("fonts").orEmpty()
            val grouped = HashMap<String, MutableList<String>>()
            for (name in names) {
                if (!name.endsWith(".ttf") && !name.endsWith(".otf")) continue
                val stem = name.substringBeforeLast('.')
                val parts = stem.split('-')
                val family = parts.first()
                val weight = if (parts.size > 1) parts[1].lowercase() else "regular"
                grouped.getOrPut(family) { mutableListOf() }.add(weight)
            }
            grouped.map { (family, weights) ->
                mapOf("family" to family, "weights" to weights)
            } + mapOf("family" to "system", "weights" to listOf("regular"))
        } catch (t: Throwable) {
            Log.w(TAG, "fonts() failed: ${t.message}")
            listOf(mapOf("family" to "system", "weights" to listOf("regular")))
        }
    }

    private fun stub(id: String, label: String): Map<String, Any?> =
        mapOf("id" to id, "label" to label, "lutPath" to "")
}

File: ios/Classes/ImgEdit/AssetCatalog.swift
import Foundation
import UIKit

/// Enumerates bundled LUTs, sticker packs, and fonts. Looks under
/// `ios/Classes/Assets/luts/`, `assets/stickers/`, and the registered
/// font family list (CTFontManagerCopyAvailableFontFamilyNames).
///
/// The bundle is resolved by the plugin class. Returns safe defaults
/// when no resources are found.
public enum AssetCatalog {

    public static func filters() -> [[String: Any]] {
        var out: [[String: Any]] = [
            ["id": "neutral", "label": "Neutral", "lutPath": ""],
        ]
        let urls = bundle().urls(forResourcesWithExtension: "cube",
                                 subdirectory: "Assets/luts") ?? []
        for url in urls {
            let id = url.deletingPathExtension().lastPathComponent
            out.append([
                "id": id,
                "label": id.capitalized,
                "lutPath": url.path,
            ])
        }
        return out
    }

    public static func stickerPacks() -> [[String: Any]] {
        let root = bundle().resourceURL?.appendingPathComponent("Assets/stickers")
        guard let root = root,
              let packs = try? FileManager.default
                .contentsOfDirectory(at: root, includingPropertiesForKeys: nil)
        else { return [] }
        var out: [[String: Any]] = []
        for pack in packs {
            let manifestURL = pack.appendingPathComponent("manifest.json")
            var label = pack.lastPathComponent
            var items: [[String: Any]] = []
            if let data = try? Data(contentsOf: manifestURL),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                label = (json["label"] as? String) ?? label
                if let arr = json["items"] as? [[String: Any]] {
                    for item in arr {
                        var copy = item
                        if let file = item["file"] as? String {
                            copy["path"] = pack.appendingPathComponent(file).path
                        }
                        items.append(copy)
                    }
                }
            }
            out.append([
                "packId": pack.lastPathComponent,
                "displayName": label,
                "items": items,
            ])
        }
        return out
    }

    public static func fonts() -> [[String: Any]] {
        var families: Set<String> = ["system"]
        let urls = bundle().urls(forResourcesWithExtension: "ttf",
                                 subdirectory: "Assets/fonts") ?? []
        for url in urls {
            let stem = url.deletingPathExtension().lastPathComponent
            let family = stem.split(separator: "-").first.map(String.init) ?? stem
            families.insert(family)
        }
        return families.sorted().map { ["family": $0, "weights": ["regular"]] }
    }

    private static func bundle() -> Bundle {
        return Bundle(for: BundleLocator.self)
    }
}

private final class BundleLocator: NSObject {}
